File: sorting/bucket.c
#include <stdint.h>
#include <stdio.h>

#include "array.h"

// function to find the max value in the array
int getMax(Arr_T A) {
  int max = A.arr[0];
  for (int i = 0; i < A.size; i++) {
    if (A.arr[i] > max) {
      max = A.arr[i];
    }
  }
  return max;
}

void bucketSort(Arr_T A) {
  int max = getMax(A); // get max of array
  int bucket[max];     // create max number of buckets
  for (int i = 0; i <= max; i++) {
    bucket[i] = 0; // fill list with 0's
  }
  for (int j = 0; j < A.size; j++) {
    bucket[A.arr[j]]++;
  }
  for (int k = 0, m = 0; k <= max; k++) {
    while (bucket[k] > 0) {
      A.arr[m++] = k;
      bucket[k]--;
    }
  }
}

int main(int argc, char *argv[]) {
  int arr_size;
  if (argc > 1) {
    if (sscanf(argv[1], "%i", &arr_size) != 1) {
      fprintf(stderr, "[ERR] - Not an integer!");
      return 1;
    }
  } else {
    arr_size = 100;
  }
  Arr_T M = make_Arr(arr_size);
  populate_Arr(M);
  printf("Array before sorting: \n");
  print_Arr(M);
  bucketSort(M);
  printf("Array after sorting: \n");
  print_Arr(M);
  return 0;
}

File: sorting/bucket.py
import array

def get_max(lst):
    max_ind = 0
    for i in range(lst.size):
        if lst.arr[i] > lst.arr[max_ind]:
            max_ind = i
    return lst.arr[max_ind]


def bucket_sort(lst):
    mx = get_max(lst)
    bucket = []
    # Fill bucket list w/ 0s
    for i in range(mx):
        bucket.append(0)
    for j in range(lst.size):
        bucket[lst.arr[j] - 1] += 1
    k = 0
    m = 0
    while k < mx:
        while bucket[k] > 0:
            lst.arr[m] = k
            m += 1
            bucket[k] -= 1
        k += 1


if __name__ == "__main__":
    array = array.Arr_T()
    array.make_Arr(100)
    array.populate_Arr()
    print("Array before sorting: ")
    array.print_Arr()
    bucket_sort(array)
    print("Array after sorting: ")
    array.print_Arr()